src/chain/shared_state.rs

```diff
@@ -0,0 +1,174 @@
+// permissions and limitations relating to use of the SAFE Network Software.
+
+use super::{ProofSet, ProvingSection, SectionInfo};
+use crate::{error::RoutingError, routing_table::DEFAULT_PREFIX, sha3::Digest256, Prefix, XorName};
+use itertools::Itertools;
+use std::collections::BTreeSet;
+
+/// Section state that is shared among all elders of a section via Parsec consensus.
+#[derive(Debug, PartialEq, Eq)]
+pub struct SharedState {
+    /// The new self section info, that doesn't necessarily have a full set of signatures yet.
+    pub new_info: SectionInfo,
+    /// The latest few fully signed infos of our own sections, each with signatures by the previous
+    /// one. This is included in every message we relay.
+    /// This is not a `BTreeSet` just now as it is ordered according to the sequence of pushes into
+    /// it.
+    pub our_infos: Vec<(SectionInfo, ProofSet)>,
+    /// Any change (split or merge) to the section that is currently in progress.
+    pub change: SectionChange,
+    // The accumulated `SectionInfo`(self or sibling) and proofs during a split pfx change.
+    pub split_cache: Option<(SectionInfo, ProofSet)>,
+    /// The set of section info hashes that are currently merging.
+    pub merging: BTreeSet<Digest256>,
+}
+
+impl SharedState {
+    pub fn new(section_info: SectionInfo) -> Self {
+        Self {
+            new_info: section_info.clone(),
+            our_infos: vec![(section_info, Default::default())],
+            change: SectionChange::None,
+            split_cache: None,
+            merging: Default::default(),
+        }
+    }
+
+    pub fn our_infos(
+        &self,
+    ) -> impl Iterator<Item = &SectionInfo> + ExactSizeIterator + DoubleEndedIterator {
+        self.our_infos.iter().map(|(si, _)| si)
+    }
+
+    pub fn opt_our_info(&self) -> Option<&SectionInfo> {
+        self.our_infos.last().map(|(si, _)| si)
+    }
+
+    /// Returns our own current section info.
+    pub fn our_info(&self) -> &SectionInfo {
+        // TODO: Replace `our_infos` with a new `NonemptyVec` type that statically guarantees that
+        // it's never empty.
+        unwrap!(self.opt_our_info())
+    }
+
+    pub fn our_prefix(&self) -> &Prefix<XorName> {
+        self.opt_our_info()
+            .map_or(&DEFAULT_PREFIX, SectionInfo::prefix)
+    }
+
+    pub fn our_version(&self) -> u64 {
+        self.opt_our_info().map_or(0, |si| *si.version())
+    }
+
+    /// Returns our section info with the given hash, if it exists.
+    pub fn our_info_by_hash(&self, hash: &Digest256) -> Option<&SectionInfo> {
+        self.our_infos
+            .iter()
+            .find(|(sec_info, _)| sec_info.hash() == hash)
+            .map(|(sec_info, _)| sec_info)
+    }
+
+    /// Returns the section info matching our own name with the given version number.
+    pub fn our_info_by_version(&self, version: u64) -> Option<&SectionInfo> {
+        // TODO: Binary search? Reverse order? Benchmark which one is fastest in practice.
```

Comment:
Do we need that TODO, either it will show in profiling as an issue and we can fix it or it won't and we don't want to do anything. (Can be follow up commit because it is copied).

Reply:
It was there before so didn't want to lose it in case we want to address it at some point (which is out of scope of this PR).

Reply:
I was thinking ok, but since this code is moved, I'd rather fix that in this PR. Unless you think this TODO is valid. (I'd rather avoid getting Performance fix that are not needed).

Reply:
I was thinking ok, but since this code is moved, I'd rather fix that in this PR. Unless you think this TODO is valid. (I'd rather avoid getting Performance fix that are not needed).

@Viv-Rajkumar could you comment?

Reply:
I'm against addressing the comment in this PR, as it can be hugely out of scope (as to do it properly it would require profiling and/or benchmarking which is not trivial). I'm OK with just deleting the comment. We should have a separate performance pass for routing at some point anyway.

Reply:
Yes, I did not mean to ask you to address the comment, only remove it.